### PR TITLE
Validate ObjectMeta in BeforeCreate

### PR DESCRIPTION
**Pull request by liggitt**

### CODE REVIEW

**Comment by liggitt** [comment]
@caesarxuchao @lavalamp PTAL

**Comment by liggitt** [comment]
at least three objects were missing name validation (events, thirdpartyresourcedata, thirdpartyresource) that this caught

**Comment by caesarxuchao** [comment]
SGTM.

**Comment by caesarxuchao** [comment]
e2e is a flake
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:912
Oct 20 15:06:57.178: updating image did not reset the back-off value in pod=pod-back-off-image/back-off d3=52s d2=21s d1=52s
```

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by lavalamp** [comment]
@liggitt I just have one question about this, which is: were clients sending invalid stuff, and if yes, who are we breaking when we turn this on?

File: pkg/api/rest/create.go [hunk 1]
```diff
@@ -68,6 +69,14 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx api.Context, obj runtime.Obje
 	if errs := strategy.Validate(ctx, obj); len(errs) > 0 {
 		return errors.NewInvalid(kind, objectMeta.Name, errs)
 	}
+
+	// Custom validation (including name validation) passed
+	// Now run common validation on object meta
+	// Do this *after* custom validation so that specific error messages are shown whenever possible
+	if errs := validation.ValidateObjectMeta(objectMeta, strategy.NamespaceScoped(), validation.ValidatePathSegmentName); len(errs) > 0 {
```
Comment (thockin) [inline pkg/api/rest/create.go]: We can be simpler, I think.  Every single object has a name that is DNS1123Subdomain or stricter.  You should just be able to pass pkg/api/validation.go:NameIsDNSSubdomain() here, and get rid of this new impl below...

I'd rather we get more rigorous about validation, but we can get there in the next release.

File: pkg/storage/util.go [hunk 2]
```diff
@@ -56,6 +58,10 @@ func NamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	if err != nil {
 		return "", err
 	}
+	name := meta.Name()
+	if ok, msg := validation.ValidatePathSegmentName(name, false); !ok {
```
Comment (thockin) [inline pkg/storage/util.go]: All of these can be replaced with pkg/util/validation.IsDNS1123Subdomain(name) (or a local wrapper thereof) and should probably be commented as "sanity checks".

File: pkg/api/rest/create.go (hunk 1 above)
Comment (liggitt) [inline pkg/api/rest/create.go]: That's not the case with downstream objects in OpenShift (like users or identities) or third party objects (the semantics of whose names are not known here). This is the minimal validation I could come up with for names used as path segments

**Comment by thockin** [comment]
This is a good safety net.  We should backport to 1.0 and 1.1 branches.

File: pkg/storage/util.go (hunk 2 above)
Comment (liggitt) [inline pkg/storage/util.go]: Object strategies should definitely validate names themselves (like most are already doing). This should be the minimal validation needed to ensure the REST and storage layers can address the stored name

Comment (thockin) [inline pkg/storage/util.go]: My point is that your PathSegmentName is redundant with DNS1123Subdomain, which already exists.  Less code is more better

Comment (liggitt) [inline pkg/storage/util.go]: DNS1123Subdomain is too strict here. There can be valid addressable object names that are not subdomains

File: pkg/api/rest/create.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/api/rest/create.go]: I think @liggitt's point is valid, especially for third party objects, I think it's better to keep the flexibility.

File: pkg/storage/util.go (hunk 2 above)
Comment (thockin) [inline pkg/storage/util.go]: hypothetically.  I'm saying I'm OK asserting that any objects we create will be a subdomain or stricter.  But it's not a huge deal, I guess.

File: pkg/api/validation/name.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+// NameMayNotBe specifies strings that cannot be used as names specified as path segments (like the REST API or etcd store)
+var NameMayNotBe = []string{".", ".."}
+
+// NameMayNotContain specifies substrings that cannot be used in names specified as path segments (like the REST API or etcd store)
+var NameMayNotContain = []string{"/", "%"}
+
+// ValidatePathSegmentName validates the name can be used as a path segment
+func ValidatePathSegmentName(name string, prefix bool) (bool, string) {
```
Comment (caesarxuchao) [inline pkg/api/validation/name.go]: prefix is not used?

**Comment by caesarxuchao** [comment]
LGTM except a nit

File: pkg/api/validation/name.go (hunk 3 above)
Comment (liggitt) [inline pkg/api/validation/name.go]: none of the places that made use of this method were passing prefix=true, but in the interest of supporting the `ValidateNameFunc` interface, I updated the impl to do the right thing if prefix=true (e.g. skip the exact name checks and only do the contains checks). Also added unit tests

Comment (caesarxuchao) [inline pkg/api/validation/name.go]: Thanks. LGTM.

**Pull request merged**